src/algebra/lie/weights.lean

```diff
@@ -0,0 +1,253 @@
+end
+
+variables {M}
+
+lemma lie_mem_pre_weight_space_of_mem_pre_weight_space {χ₁ χ₂ : L → R} {x : L} {m : M}
+  (hx : x ∈ pre_weight_space L χ₁) (hm : m ∈ pre_weight_space M χ₂) :
+  ⁅x, m⁆ ∈ pre_weight_space M (χ₁ + χ₂) :=
+begin
+  apply weight_vector_multiplication L M M (to_module_hom R L M) χ₁ χ₂,
+  simp only [lie_module_hom.coe_to_linear_map, function.comp_app, linear_map.coe_comp,
+    tensor_product.map_incl, linear_map.mem_range],
+  use [⟨x, hx⟩ ⊗ₜ ⟨m, hm⟩],
+  simp only [submodule.subtype_apply, to_module_hom_apply, tensor_product.map_tmul],
+  refl,
+end
+
+variables (M)
+
+/-- If a Lie algebra is nilpotent, then pre-weight spaces are Lie submodules. -/
+def weight_space [lie_algebra.is_nilpotent R L] (χ : L → R) : lie_submodule R L M :=
+{ lie_mem := λ x m hm,
+  begin
+    rw ← zero_add χ,
+    refine lie_mem_pre_weight_space_of_mem_pre_weight_space _ hm,
+    suffices : pre_weight_space L (0 : L → R) = ⊤, { simp only [this, submodule.mem_top], },
+    exact lie_algebra.infi_max_gen_zero_eigenspace_eq_top_of_nilpotent R L,
+  end,
+  .. pre_weight_space M χ }
+
+@[simp] lemma mem_weight_space [lie_algebra.is_nilpotent R L] (χ : L → R) (m : M) :
+  m ∈ weight_space M χ ↔ m ∈ pre_weight_space M χ :=
+iff.rfl
+
+/-- See also the more useful form `lie_module.zero_weight_space_eq_top_of_nilpotent`. -/
+@[simp] lemma zero_weight_space_eq_top_of_nilpotent'
+  [lie_algebra.is_nilpotent R L] [is_nilpotent R L M] :
+  weight_space M (0 : L → R) = ⊤ :=
+begin
+  rw [← lie_submodule.coe_to_submodule_eq_iff, lie_submodule.top_coe_submodule],
+  exact infi_max_gen_zero_eigenspace_eq_top_of_nilpotent R L M,
+end
+
+@[simp] lemma coe_weight_space_of_top [lie_algebra.is_nilpotent R L] (χ : L → R) :
+  (weight_space M (χ ∘ (⊤ : lie_subalgebra R L).incl) : submodule R M) = weight_space M χ :=
+begin
+  ext m,
+  simp only [weight_space, lie_submodule.coe_to_submodule_mk, lie_subalgebra.coe_bracket_of_module,
+    function.comp_app, mem_pre_weight_space],
+  split; intros h x,
+  { obtain ⟨k, hk⟩ := h ⟨x, set.mem_univ x⟩, use k, exact hk, },
+  { obtain ⟨k, hk⟩ := h x, use k, exact hk, },
+end
+
+@[simp] lemma zero_weight_space_eq_top_of_nilpotent
+  [lie_algebra.is_nilpotent R L] [is_nilpotent R L M] :
+  weight_space M (0 : (⊤ : lie_subalgebra R L) → R) = ⊤ :=
+begin
+  /- We use `coe_weight_space_of_top` as a trick to circumvent the fact that we don't (yet) know
+    `is_nilpotent R (⊤ : lie_subalgebra R L) M` is equivalent to `is_nilpotent R L M`. -/
+  have h₀ : (0 : L → R) ∘ (⊤ : lie_subalgebra R L).incl = 0, { ext, refl, },
+  rw [← lie_submodule.coe_to_submodule_eq_iff, lie_submodule.top_coe_submodule, ← h₀,
+    coe_weight_space_of_top, ← infi_max_gen_zero_eigenspace_eq_top_of_nilpotent R L M],
+  refl,
+end
+
+/-- Given a Lie module `M` of a Lie algebra `L`, a weight of `M` with respect to a nilpotent
+subalgebra `H ⊆ L` is a Lie character whose corresponding weight space is non-empty. -/
+structure weight :=
```

Comment:
My intuition says that most of the time we might want to work with an unbundled `is_weight` predicate on characters. And only at a later stage we bundle them into the weight lattice.

Reply:
Agreed; will unbundle!